File: src/main/java/com/fouan/zones/view/ComputedZones.java
package com.fouan.zones.view;

import com.fouan.actors.ActorId;
import com.fouan.zones.Position;
import com.fouan.zones.Zone;
import lombok.Getter;

import java.util.*;

public final class ComputedZones {

    private final Map<Position, ComputedZone> zones;

    public ComputedZones() {
        zones = new HashMap<>();
    }

    public Collection<ComputedZone> all() {
        return zones.values();
    }

    public void add(ComputedZone zone) {
        if (zones.containsKey(zone.getPosition())) {
            throw new IllegalStateException("Zone already exists");
        }

        zones.put(zone.getPosition(), zone);
    }

    public void update(Position position, ComputedZone zone) {
        if (!zones.containsKey(zone.getPosition())) {
            throw new IllegalStateException("Zone does not exist");
        }

        zones.put(position, zone);
    }

    public Optional<ComputedZone> findByPosition(Position position) {
        return Optional.ofNullable(zones.get(position));
    }

    @Getter
    public static class ComputedZone {
        private final Zone zone;

        private final List<ActorId> actorIds;

        public ComputedZone(Zone zone) {
            this(zone, Collections.emptyList());
        }

        public ComputedZone(Zone zone, List<ActorId> actorIds) {
            this.zone = zone;
            this.actorIds = actorIds;
        }

        public Position getPosition() {
            return zone.getPosition();
        }
    }
}

File: src/main/java/com/fouan/zones/view/ComputedZones.kt
package com.fouan.zones.view

import com.fouan.actors.ActorId
import com.fouan.zones.Position
import com.fouan.zones.Zone
import java.util.function.UnaryOperator

class ComputedZones {
    private val zones: MutableMap<Position, ComputedZone> = mutableMapOf()

    fun all(): Collection<ComputedZone> {
        return zones.values
    }

    fun add(zone: ComputedZone) {
        check(!zones.containsKey(zone.position)) { "Zone already exists" }
        zones[zone.position] = zone
    }

    fun update(position: Position, updater: UnaryOperator<ComputedZone>) {
        findByPosition(position)
            ?.let { zones[position] = updater.apply(it) }
    }

    fun findByPosition(position: Position): ComputedZone? {
        return zones[position]
    }

    fun findByActorId(actorId: ActorId): ComputedZone? {
        return zones.values
            .find { it.actorIds.contains(actorId) }
    }

    class ComputedZone(
        val zone: Zone,
        val actorIds: List<ActorId> = emptyList()
    ) {
        val noiseTokens = 0
        val position: Position
            get() = zone.position
    }
}
